src/KubernetesClient/Watcher.cs

```diff
@@ -174,6 +150,70 @@ private async Task WatcherLoop(CancellationToken cancellationToken)
             }
         }
 
+        internal static async IAsyncEnumerable<(WatchEventType, T)> CreateWatchEventEnumerator(
+            Func<Task<TextReader>> streamReaderCreator,
+            Action<Exception> onError = null,
+            [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            Task<TR> AttachCancellationToken<TR>(Task<TR> task)
+            {
+                if (!task.IsCompleted)
+                {
+                    // here to pass cancellationToken into task
+                    return task.ContinueWith(t => t.GetAwaiter().GetResult(), cancellationToken);
+                }
+
+                return task;
+            }
+
+            using var streamReader = await AttachCancellationToken(streamReaderCreator()).ConfigureAwait(false);
+
+            for (; ; )
+            {
+                // ReadLineAsync will return null when we've reached the end of the stream.
+                var line = await AttachCancellationToken(streamReader.ReadLineAsync()).ConfigureAwait(false);
+
+                if (cancellationToken.IsCancellationRequested)
```

Comment:
Typically better to use `cancellationToken.ThrowIfCancellationRequested()`

Reply:
agree 
but here is to be compatible with old watcher pattern
otherwise, signal a cancel will raise an exception.

Reply:
For callers using `await foreach(...WatchAsync(CT))` it could be expected when a cancel is requested the await will throw? Might not be a mandatory convention, but it would indicate if the loop ended normally from response ending vs ending from a client-side cancellation teardown.

Ah - also the await keyword in the loop will also throw the exception if the ReadLineAsync task enters a cancelled state, so you'll need to catch for that as well for old watcher pattern.

How about if the `WatcherLoop` method has an additional `catch (OperationCancelledException)` case which does not call OnError? 

That would stop cancel exceptions from reaching the old pattern, but enable you to use `ThrowIfCancellationRequested` from the async enumerable function and allow `await foreach` callers to have task.Canceled status propagate back.